File: MyFarmProduce/Controllers/ProfileController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MyFarmProduce.Application.Interfaces;
using MyFarmProduce.Common;
using MyFarmProduce.Web.Extensions;
using MyFarmProduce.Web.Models;

namespace MyFarmProduce.Web.Controllers;

// Customer profile: edit name/photo; phone & email are locked (change via admin request).
[Authorize(Roles = AppConstants.Roles.Customer)]
public class ProfileController : Controller
{
    private readonly IProfileService _profiles;
    private readonly IFileStorage _files;

    public ProfileController(IProfileService profiles, IFileStorage files)
    {
        _profiles = profiles;
        _files = files;
    }

    [HttpGet]
    public async Task<IActionResult> Index()
    {
        var customer = await _profiles.GetCustomerAsync(User.GetCustomerId());
        if (customer is null) return NotFound();

        return View(new CustomerProfileViewModel
        {
            Customer = customer,
            Name = customer.Name,
            ChangeRequests = await _profiles.GetMyChangeRequestsAsync(customer.Id)
        });
    }

    [HttpPost, ValidateAntiForgeryToken]
    public async Task<IActionResult> Index(string name, IFormFile? photoFile)
    {
        string? photoUrl = null;
        if (photoFile is { Length: > 0 })
        {
            await using var stream = photoFile.OpenReadStream();
            photoUrl = await _files.SaveImageAsync(stream, photoFile.FileName, AppConstants.UploadFolders.Avatars);
        }

        await _profiles.UpdateCustomerProfileAsync(User.GetCustomerId(), name, photoUrl);
        TempData["Message"] = "Profile updated.";
        return RedirectToAction(nameof(Index));
    }

    [HttpPost, ValidateAntiForgeryToken]
    public async Task<IActionResult> RequestChange(string field, string requestedValue)
    {
        if (field is not ("Phone" or "Email") || string.IsNullOrWhiteSpace(requestedValue))
        {
            TempData["Message"] = "Please provide a valid value.";
            return RedirectToAction(nameof(Index));
        }

        await _profiles.RequestFieldChangeAsync(User.GetCustomerId(), field, requestedValue);
        TempData["Message"] = $"Request to change your {field.ToLower()} was submitted for admin approval.";
        return RedirectToAction(nameof(Index));
    }
}

File: web/src/main/java/com/myfarmproduce/web/controller/ProfileController.java
package com.myfarmproduce.web.controller;

import com.myfarmproduce.application.service.FileStorage;
import com.myfarmproduce.application.service.ProfileService;
import com.myfarmproduce.common.AppConstants;
import com.myfarmproduce.web.security.AppPrincipal;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import java.io.IOException;
import java.io.UncheckedIOException;

@Controller
@RequestMapping("/profile")
public class ProfileController {

    private final ProfileService profiles;
    private final FileStorage files;

    public ProfileController(ProfileService profiles, FileStorage files) {
        this.profiles = profiles;
        this.files = files;
    }

    @GetMapping
    public String index(@AuthenticationPrincipal AppPrincipal principal, Model model, HttpServletResponse response) {
        var customer = profiles.getCustomer(principal.getId());
        if (customer.isEmpty()) { response.setStatus(404); return "error/404"; }

        model.addAttribute("customer", customer.get());
        model.addAttribute("changeRequests", profiles.getMyChangeRequests(customer.get().getId()));
        return "profile/index";
    }

    @PostMapping
    public String update(@RequestParam String name, @RequestParam(required = false) MultipartFile photoFile,
                          @AuthenticationPrincipal AppPrincipal principal, RedirectAttributes redirectAttributes) {
        String photoUrl = saveUpload(photoFile, AppConstants.UploadFolders.AVATARS);
        profiles.updateCustomerProfile(principal.getId(), name, photoUrl);
        redirectAttributes.addFlashAttribute("message", "Profile updated.");
        return "redirect:/profile";
    }

    @PostMapping("/request-change")
    public String requestChange(@RequestParam String field, @RequestParam String requestedValue,
                                 @AuthenticationPrincipal AppPrincipal principal, RedirectAttributes redirectAttributes) {
        if (!("Phone".equals(field) || "Email".equals(field)) || !StringUtils.hasText(requestedValue)) {
            redirectAttributes.addFlashAttribute("message", "Please provide a valid value.");
            return "redirect:/profile";
        }

        profiles.requestFieldChange(principal.getId(), field, requestedValue);
        redirectAttributes.addFlashAttribute("message",
                "Request to change your " + field.toLowerCase() + " was submitted for admin approval.");
        return "redirect:/profile";
    }

    private String saveUpload(MultipartFile file, String folder) {
        if (file == null || file.isEmpty()) return null;
        try {
            return files.saveImage(file.getInputStream(), file.getOriginalFilename(), folder);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
